LOOP-6/LOOP-4 Implement TidepoolKit upload of initial data types

darinkrauss opened the pull request:
- https://tidepool.atlassian.net/browse/LOOP-4
- https://tidepool.atlassian.net/browse/LOOP-6
- Get profile, list/create data sets, list/create data APIs
- Use Tidepool API compliant JSON encoding and decoding
- Defacto encoding and decoding for generic dictionary from JSON
- Implement base Tidepool API datum
- Initial CGM and Food datum types
- Add association, dictionary, location, and origin common structs
- Add blood glucose constants and target structs
- Add profile and contained patient structs
- Add dataset struct
- Updates to example app to reflect TAPI changes
- Add sharing to example app for debugging purposes
- Add sample data
- Refactor errors to capture all specific conditions
- Refactor request create and perform request to utilize Codable
- Replace all usage RawRepresentation with Codable
- Loads of tests

darinkrauss committed:
LOOP-6/LOOP-4 Implement TidepoolKit upload of initial data types

- LOOP-6 Implement TidepoolKit data upload
- LOOP-4 Implement initial TidepoolKit data types
- Get profile, list/create data sets, list/create data APIs
- Use Tidepool API compliant JSON encoding and decoding
- Defacto encoding and decoding for generic dictionary from JSON
- Implement base Tidepool API datum
- Initial CGM and Food datum types
- Add association, dictionary, location, and origin common structs
- Add blood glucose constants and target structs
- Add profile and contained patient structs
- Add dataset struct
- Updates to example app to reflect TAPI changes
- Add sharing to example app for debugging purposes
- Add sample data
- Refactor errors to capture all specific conditions
- Refactor request create and perform request to utilize Codable
- Replace all usage RawRepresentation with Codable
- Loads of tests

rickpasetto commented on TidepoolKit/Internal/Extensions/Codable.swift:
```diff
@@ -0,0 +1,222 @@
+    }
+}
+
+extension KeyedDecodingContainer {
```
Maybe I just cannot find them, but where are the unit tests for these extensions?

darinkrauss replied:
Frankly, I'm not a fan of testing internal implementation details, within which this is included. This is akin to testing `private` functions and properties on a struct/class.

In general, my feeling is that if the public-facing "API" is fully and properly tested then the internal implementation should also be tested.

However, if you feel strongly about unit tests here I will implement them. @rickpasetto

rickpasetto replied:
I do feel strongly about it, @darinkrauss ,  but it shouldn't block your PR.  We should discuss it in another forum.

rickpasetto commented on TidepoolKit/Internal/Extensions/Date.swift:
```diff
@@ -0,0 +1,70 @@
+        return Date.deviceTimeFormatter.string(from: self.roundedToTimeInterval(TimeInterval.millisecond))
+    }
+
+    static let timeFormatter: ISO8601DateFormatter = {
```
[nit] This is fine, but I was just wondering, are these alternate time formats associated with specific versions of the Tidepool API?  If so, it might be handy to either name these as such or maybe just comment.

darinkrauss replied:
These two Date formats cover all of the Tidepool data model and are not version dependent. However, I just noticed the extension is `public` and need not be. I will change to avoid cluttering the public namespace.

rickpasetto commented on TidepoolKit/Internal/Extensions/Date.swift:
```diff
@@ -0,0 +1,70 @@
+
+import Foundation
+
+public extension Date {
```
Unit tests for these?

darinkrauss replied:
I removed the accidental inclusion of the `public` attribute on this extension. I did not provide unit tests per not testing internal implementation details, as above.

rickpasetto commented on TidepoolKit/Internal/Responses/DataResponse.swift:
```diff
@@ -0,0 +1,55 @@
+                case .food:
+                    datum = try TFoodDatum(from: superDecoder)
+                default:
+                    break // DEPRECATED: Ignore
```
Seems if this is not an error, we should probably at least log.  If, someday, the backend adds a type, and someone has an old version of the app installed, there'd be no indication of this.  My recommendation, though, is to throw an error in this case.

darinkrauss replied:
Sorry, the comment is misleading. The default case is only because I did not include all of the implemented Tidepool data model types in this PR. Once all types are included (in subsequent PRs) there will be no default case and any unexpected type (aka a new or renamed type) will throw an error when decoding the `enum`. The thrown error will be captured and the "bad" data will be added to the malformed list (unparsable for whatever reason) that can be logged by the client.

rickpasetto commented on TidepoolKit/Internal/Responses/DataResponse.swift:
```diff
@@ -0,0 +1,55 @@
+        self.malformed = malformed
+    }
+    
+    init(from decoder: Decoder) throws {
```
Unit tests?

darinkrauss replied:
I did not provide unit tests per not testing internal implementation details, as above.

rickpasetto commented on TidepoolKit/Internal/Responses/DataResponse.swift:
```diff
@@ -0,0 +1,55 @@
+//
+
+struct DataResponse: Codable {
+    var data: [TDatum]
```
General comment about all of the `Codable` model objects: I notice you use `var`s for their members...is it your intent that they should be mutable like this?  I'm curious why they are not all `let`s.

darinkrauss replied:
Good question. 

One thing I struggled with was dealing with all (17) of the properties common to all Tidepool data types (as implemented in the TDatum base class). If I implemented properties as `let` then every single property would have to be specified in every initializer for all Tidepool data types. This got very unreadable and annoying very fast. 

I decided the ability to create the object using a straightforward, type-specific constructor (with only the few type-specific properties) and then allowing modification of the objects afterwards (from `var`) to set any of the infrequently-used common properties made the most sense.

In addition, while writing code to use these classes, I realized that there is the added benefit of being able to set these common properties en-mass after creating a number of data.

While it is not ideal I think this is a reasonable compromise.

rickpasetto replied:
I admit that adding all those initializers is a drag.  In the past, I've done exactly what you've done: started with `var`s because it was such a pain.  In the end, I eventually came around and realized that immutability is actually more beneficial to the users, and it was just boilerplate code anyway.  That said, not a blocker for this review, but something to chew on...

rickpasetto commented on TidepoolKit/TAPI.swift:
```diff
@@ -84,25 +83,25 @@ public class TAPI {
+                    completion(.failure(error))
+                }
+            case .success(let response, let data, let loginResponse):
+                if let authenticationToken = response.value(forHTTPHeaderField: "X-Tidepool-Session-Token"), !authenticationToken.isEmpty {
```
[nit] consider putting this header key into a constant

rickpasetto commented on TidepoolKit/TAPI.swift:
```diff
@@ -145,40 +144,199 @@ public class TAPI {
+        request?.setValue(session.authenticationToken, forHTTPHeaderField: "X-Tidepool-Session-Token")
         if let trace = session.trace {
-            request.setValue(trace, forHTTPHeaderField: "X-Tidepool-Trace-Session")
+            request?.setValue(trace, forHTTPHeaderField: "X-Tidepool-Trace-Session")
```
Similar comment regarding a constant for header field key

rickpasetto commented on TidepoolKit/TAPI.swift:
```diff
@@ -145,40 +144,199 @@ public class TAPI {
+
+    private func createRequest<E>(session: TSession, method: String, path: String, body: E) -> URLRequest? where E: Encodable {
+        var request = createRequest(session: session, method: method, path: path)
+        request?.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
```
consider putting `contentType` as a constant

rickpasetto reviewed (commented):
🌟 for committing to `Codable` !  Good job.

Some general thoughts and comments:
1. Great job on the unit tests! I also would love to see some tests on the extensions and some other smaller classes you created (comments were added inline)
2. I'm not sure why Model objects need `var` members.  Prefer immutable objects where you can.
3. You've had to implement a lot of HTTP request/response code on your own here.  Did you consider using a library such as [AlamoFire](https://github.com/Alamofire/Alamofire) to help?  (it would save you from having to write unit tests for all if it too).

Oh one last thing: consider enabling unit test code coverage in the test scheme.

nhamming reviewed (approved):
Lots of code, but nothing jumped out at me.

darinkrauss committed:
Updates based upon PR review

- Remove public attribute on Date extension
- Use constants for HTTP header field names

rickpasetto reviewed (approved):
LGTM 👍
